File: src/main/java/net/guizhanss/fastmachines/libs/guizhanlib/utils/FileUtil.java
package net.guizhanss.fastmachines.libs.guizhanlib.utils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.BiPredicate;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

/**
 * Java-8-safe port of the small slice of GuizhanLib's {@code FileUtil} this addon uses.
 */
public final class FileUtil {

    private FileUtil() {
    }

    public static <T> List<T> listJarEntries(File jarFile, BiPredicate<String, JarEntry> filter,
                                              BiFunction<String, JarEntry, T> mapper) throws IOException {
        List<T> result = new ArrayList<>();
        try (JarFile jar = new JarFile(jarFile)) {
            Enumeration<JarEntry> entries = jar.entries();
            while (entries.hasMoreElements()) {
                JarEntry entry = entries.nextElement();
                if (filter.test(entry.getName(), entry)) {
                    result.add(mapper.apply(entry.getName(), entry));
                }
            }
        }
        return result;
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/libs/guizhanlib/utils/FileUtil.kt
package net.guizhanss.fastmachines.libs.guizhanlib.utils

import java.io.File
import java.util.jar.JarEntry
import java.util.jar.JarFile

/**
 * Java-8-safe, Kotlin port of the small slice of GuizhanLib's `FileUtil` this addon uses - see the
 * package-level note in [net.guizhanss.fastmachines.libs.guizhanlib].
 */
object FileUtil {

    fun <T> listJarEntries(
        jarFile: File,
        filter: (String, JarEntry) -> Boolean,
        mapper: (String, JarEntry) -> T,
    ): List<T> {
        JarFile(jarFile).use { jar ->
            return jar.entries().asSequence()
                .filter { filter(it.name, it) }
                .map { mapper(it.name, it) }
                .toList()
        }
    }
}
